cirq/experiments/single_qubit_readout_calibration.py

```diff
@@ -0,0 +1,112 @@
+
+import dataclasses
+import time
+
+import numpy as np
+
+from cirq import circuits, ops, work
+
+
+@dataclasses.dataclass
+class SingleQubitReadoutCalibrationResult:
+    """Result of estimating single qubit readout error.
+
+    Attributes:
+        zero_state_errors: A dictionary from qubit to probability of measuring
+            a 1 when the qubit is initialized to |0⟩.
+        one_state_errors: A dictionary from qubit to probability of measuring
+            a 0 when the qubit is initialized to |1⟩.
+        repetitions: The number of repetitions that were used to estimate the
+            probabilities.
+        timestamp: A timestamp for the time the data was taken.
+    """
+    zero_state_errors: Dict[ops.Qid, float]
+    one_state_errors: Dict[ops.Qid, float]
+    repetitions: int
+    timestamp: float
+
+    def _json_dict_(self):
+        return {
+            'cirq_type': self.__class__.__name__,
+            'zero_state_errors': list(self.zero_state_errors.items()),
+            'one_state_errors': list(self.one_state_errors.items()),
+            'repetitions': self.repetitions,
+            'timestamp': self.timestamp
+        }
+
+    @classmethod
+    def _from_json_dict_(cls, zero_state_errors, one_state_errors, repetitions,
+                         timestamp, **kwargs):
+        return cls(zero_state_errors=dict(zero_state_errors),
+                   one_state_errors=dict(one_state_errors),
+                   repetitions=repetitions,
+                   timestamp=timestamp)
+
+    def __repr__(self):
+        return ('cirq.experiments.SingleQubitReadoutCalibrationResult('
+                f'zero_state_errors={self.zero_state_errors!r}, '
+                f'one_state_errors={self.one_state_errors!r}, '
+                f'repetitions={self.repetitions!r}, '
+                f'timestamp={self.timestamp!r})')
+
+
+def estimate_single_qubit_readout_errors(
+        sampler: work.Sampler,
+        *,
+        qubits: Iterable[ops.Qid],
+        repetitions: int = 1000) -> SingleQubitReadoutCalibrationResult:
+    """Estimate single-qubit readout error.
+
+    For each qubit, prepare the |0⟩ state and measure. Calculate how often a 1
+    is measured. Also, prepare the |1⟩ state and calculate how often a 0 is
+    measured. The state preparations and measurements are done in parallel,
+    i.e., for the first experiment, we actually prepare every qubit in the |0⟩
+    state and measure them simultaneously.
+
+    Args:
+        sampler: The quantum engine or simulator to run the circuits.
+        qubits: The qubits being tested.
+        repetitions: The number of measurement repetitions to perform.
+
+    Returns:
+        A SingleQubitReadoutCalibrationResult storing the readout error
+        probabilities as well as the number of repetitions used to estimate
+        the probabilties.
+    """
+    qubits = list(qubits)
+
+    zeros_circuit = circuits.Circuit(ops.measure_each(*qubits, key_func=repr))
```

Comment:
why `repr` and not `str`?

Reply:
The repr is required to be unique for different qubits.

Reply:
I think this is also the case for `str`, no?